core/arch/riscv64/mangle.c

```diff
@@ -385,13 +394,406 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                                                    opnd_get_disp(curop),
+                                                    opnd_get_size(curop)));
+            }
+        }
+        instr_set_translation(instr, instrlist_get_translation_target(ilist));
+
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(dcontext, scratch_reg,
+                                     os_get_app_tls_base_offset(TLS_REG_LIB)));
+        PRE(ilist, next_instr,
+            instr_create_restore_from_tls(dcontext, scratch_reg, slot));
+    }
+
+    /* If instr uses stolen register, we use it from app's TLS through a scratch register.
+     *
+     * TODO i#3544: If stolen register is only used for src, do not spill back into app's
+     * TLS; likewise, if it's only used for dst, do not restore it from app's TLS.
+     */
+    if (instr_uses_reg(instr, dr_reg_stolen)) {
+        scratch_reg = pick_scratch_reg(dcontext, instr, scratch_reg, &slot);
+        PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg, slot));
+        PRE(ilist, instr,
+            instr_create_restore_from_tls(dcontext, scratch_reg, TLS_REG_STOLEN_SLOT));
+
+        for (i = 0; i < instr_num_dsts(instr); i++) {
+            curop = instr_get_dst(instr, i);
+            if (opnd_is_reg(curop) && opnd_get_reg(curop) == dr_reg_stolen)
+                instr_set_dst(instr, i, opnd_create_reg(scratch_reg));
+            else if (opnd_is_base_disp(curop) && opnd_get_base(curop) == dr_reg_stolen) {
+                instr_set_dst(instr, i,
+                              opnd_create_base_disp(scratch_reg, DR_REG_NULL, 0,
+                                                    opnd_get_disp(curop),
+                                                    opnd_get_size(curop)));
+            }
+        }
+        for (i = 0; i < instr_num_srcs(instr); i++) {
+            curop = instr_get_src(instr, i);
+            if (opnd_is_reg(curop) && opnd_get_reg(curop) == dr_reg_stolen)
+                instr_set_src(instr, i, opnd_create_reg(scratch_reg));
+            else if (opnd_is_base_disp(curop) && opnd_get_base(curop) == dr_reg_stolen) {
+                instr_set_src(instr, i,
+                              opnd_create_base_disp(scratch_reg, DR_REG_NULL, 0,
+                                                    opnd_get_disp(curop),
+                                                    opnd_get_size(curop)));
+            }
+        }
+        instr_set_translation(instr, instrlist_get_translation_target(ilist));
+
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(dcontext, scratch_reg, TLS_REG_STOLEN_SLOT));
+        PRE(ilist, next_instr,
+            instr_create_restore_from_tls(dcontext, scratch_reg, slot));
+    }
+}
+
+static void
+mangle_cbr_stolen_reg_and_tp_reg(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                                 instr_t *next_instr)
+{
+    instr_t *fall = INSTR_CREATE_label(dcontext);
```

Comment:
Add ASSERT_NOT_TESTED(), which can be removed once we have an actual test.  That makes it clear this has not actually been testing.  (Or if you are able to run an end to end test: please include that test in this PR.)